src/VkLayer_GF_amber_scoop/src/buffer_copy.cc

```diff
@@ -0,0 +1,206 @@
+        FindMemoryType(memory_requirements.memoryTypeBits,
+                       VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    RequireSuccess(
+        device_data_->vkAllocateMemory(device, &vkMemoryAllocateInfo, nullptr,
+                                       &buffer_copy_memory_),
+        "Failed to allocate memory for buffer copy.");
+  }
+  RequireSuccess(device_data_->vkBindBufferMemory(device, buffer_copy_,
+                                                  buffer_copy_memory_, 0),
+                 "Failed binding memory for buffer copy.");
+
+  // Create command buffer.
+  VkCommandBuffer command_buffer = nullptr;
+  VkCommandBufferAllocateInfo command_buffer_allocate_info = {};
+  command_buffer_allocate_info.sType =
+      VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+  command_buffer_allocate_info.commandPool = command_pool;
+  command_buffer_allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+  command_buffer_allocate_info.commandBufferCount = 1;
+  RequireSuccess(device_data_->vkAllocateCommandBuffers(
+                     device, &command_buffer_allocate_info, &command_buffer),
+                 "Failed to allocate command buffers.");
+
+  // We have created a dispatchable object, so we must set the dispatch table
+  // pointer.
+  CopyDispatchTablePointer(device, command_buffer);
+
+  // Record to the command buffer.
+  VkCommandBufferBeginInfo begin_info = {};
+  begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+
+  RequireSuccess(
+      device_data_->vkBeginCommandBuffer(command_buffer, &begin_info),
+      "Failed to begin command buffer.");
+
+  // Memory barrier for making sure we can safely transfer the data to a host
+  // visible memory.
+  // TODO(ilkkasaa): This barrier could be submitted only before the first
+  //   buffer copy operation. Current implementation submits the same barrier
+  //   for every buffer copy operation which is redundant.
+  VkMemoryBarrier memory_barrier;
+  memory_barrier.sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER;
+  memory_barrier.pNext = nullptr;
+  memory_barrier.srcAccessMask =
+      VK_ACCESS_HOST_WRITE_BIT | VK_ACCESS_TRANSFER_WRITE_BIT;
+  memory_barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
+
+  device_data_->vkCmdPipelineBarrier(
+      command_buffer,                      // commandBuffer
+      VK_PIPELINE_STAGE_HOST_BIT |         //
+          VK_PIPELINE_STAGE_TRANSFER_BIT,  // srcStageMask
+      VK_PIPELINE_STAGE_TRANSFER_BIT,      // dstStageMask
+      0,                                   // dependencyFlags
+      1U,                                  // memoryBarrierCount
+      &memory_barrier,                     // pMemoryBarriers
+      0,                                   // bufferMemoryBarrierCount
+      nullptr,                             // pBufferMemoryBarriers
+      0,                                   // imageMemoryBarrierCount
+      nullptr);                            // pImageMemoryBarriers
```

Comment:
For the srcStageMask and srcAccessMask, I think we should probably add more bits to include things like shader writes, etc. See how Amber does it: https://github.com/google/amber/blob/9f59702d3ddcb08dbec261c598de2e82e8fb3936/src/vulkan/resource.cc#L196 (but the dstStageMask and dstAccessMask is fine I would think)

Reply:
Ah, I see.